Binary.agda

```diff
@@ -1,62 +1,86 @@
 module Binary where
+open import Data.Nat using (ℕ; _+_; _*_; suc; zero)
 
--- data Digit : Set where
---     one : Digit
---     zero : Digit
+-- representation
+data Bin : Set where
+    ⟨⟩ : Bin
+    _O : Bin -> Bin
+    _I : Bin -> Bin
 
--- data Bin : Set where
---     base : Bin
---     suc : Digit -> Bin -> Bin
+inc : Bin -> Bin
+inc ⟨⟩ = ⟨⟩ I
+inc (bin O) = bin I
+inc (bin I) = (inc bin) O
 
--- _+_ : Bin -> Bin -> Bin
--- base + b = b
--- suc x a + base = suc x a
--- suc one a + suc one b = suc one (suc zero (a + b))
--- suc one a + suc zero b = suc one (a + b)
--- suc zero a + suc one b = suc one (a + b)
--- suc zero a + suc zero b = a + b
--- -- hmmm
 
--- -- Denotation for binary arithmetic?
--- -- Bin -> Bin -> Bin
+_+ᵇ_ : Bin -> Bin -> Bin
+⟨⟩ +ᵇ b = b
+(a O) +ᵇ ⟨⟩ = (a O)
+(a I) +ᵇ ⟨⟩ = (a I)
+(a O) +ᵇ (b O) = (a +ᵇ b) O
+(a O) +ᵇ (b I) = (a +ᵇ b) I
+(a I) +ᵇ (b O) = (a +ᵇ b) I
+(a I) +ᵇ (b I) = (inc (a +ᵇ b)) O
+-- ⟨⟩ +ᵇ b = b
+-- a +ᵇ ⟨⟩ = a 
+-- (a O) +ᵇ (b O) = (a +ᵇ b) O
+-- (a O) +ᵇ (b I) = (a +ᵇ b) I
+-- (a I) +ᵇ (b O) = (a +ᵇ b) I
+-- (a I) +ᵇ (b I) = (inc (a +ᵇ b)) O
 
--- -- define a meaning function
+0ᵇ : Bin
+0ᵇ = ⟨⟩
 
--- open import Data.Nat using (ℕ)
+⟦_⟧ : Bin -> ℕ 
+⟦ ⟨⟩ ⟧ = zero
+⟦ x O ⟧ = ⟦ x ⟧ * 2
+⟦ x I ⟧ = ⟦ x ⟧ * 2 + 1
 
--- Bin : Set
--- Bin = ℕ -> Digit 
 
--- _D+_ : Bin -> Bin -> Bin
--- (f D+ g) ℕ.zero = {! f ℕ.zero + g ℕ.zero   !}
--- (f D+ g) (ℕ.suc x) = {! f (ℕ.suc x) + g (ℕ.suc x) + (f D+ g) x  !}
---     where 
---     _dig+_ : Digit -> Digit -> 
+import Relation.Binary.PropositionalEquality as Eq
+open Eq using (_≡_; refl; cong; sym)
+open Eq.≡-Reasoning using (begin_; _≡⟨⟩_; step-≡; _∎)
+open import Data.Nat.Properties using (+-identityʳ; *-distribʳ-+; +-assoc; +-comm)
+0ᴴ : ⟦ 0ᵇ ⟧ ≡ 0
+0ᴴ = refl
 
+inc-≡ : ∀ (bin : Bin) -> ⟦ inc bin ⟧ ≡ ⟦ bin ⟧ + 1
+inc-≡ ⟨⟩ = refl
+inc-≡ (bin O) = refl
+inc-≡ (bin I) 
+    rewrite inc-≡ bin 
+    | *-distribʳ-+ 2 (⟦ bin ⟧) 1
+    | +-assoc (⟦ bin ⟧ * 2) 1 1 = refl
 
-open import Data.Nat using (ℕ; _≟_; suc; zero)
-open import Data.Bool using (Bool; true; false; T; _∧_; _∨_; _xor_; not)
-open import Data.List using (List; _∷_; [])
-open import Data.List.Base using (any)
-open import Relation.Nullary.Decidable using (⌊_⌋)
 
+_+ᴴ_ : ∀ x y → ⟦ x +ᵇ y ⟧ ≡ ⟦ x ⟧ + ⟦ y ⟧
+⟨⟩ +ᴴ b = refl
+(a O) +ᴴ ⟨⟩ rewrite +-identityʳ (⟦ a ⟧ * 2) = refl
+(a I) +ᴴ ⟨⟩ rewrite +-identityʳ (⟦ a ⟧ * 2 + 1) = refl
+(a O) +ᴴ (b O) rewrite a +ᴴ b | *-distribʳ-+ 2 (⟦ a ⟧) (⟦ b ⟧)  = refl
+(a O) +ᴴ (b I) rewrite a +ᴴ b 
+    | *-distribʳ-+ 2 (⟦ a ⟧) (⟦ b ⟧)
+    | +-assoc (⟦ a ⟧ * 2) (⟦ b ⟧ * 2) 1  = refl
+(a I) +ᴴ (b O) 
+    rewrite a +ᴴ b 
+    | *-distribʳ-+ 2 (⟦ a ⟧) (⟦ b ⟧)  -- ⟦ a ⟧ * 2 + ⟦ b ⟧ * 2 + 1 ≡ ⟦ a ⟧ * 2 + 1 + ⟦ b ⟧ * 2
+    | +-assoc (⟦ a ⟧ * 2) (⟦ b ⟧ * 2) 1 -- ⟦ a ⟧ * 2 + (⟦ b ⟧ * 2 + 1) ≡ ⟦ a ⟧ * 2 + 1 + ⟦ b ⟧ * 2
+    | +-comm (⟦ b ⟧ * 2 ) 1 -- ⟦ a ⟧ * 2 + suc (⟦ b ⟧ * 2) ≡ ⟦ a ⟧ * 2 + 1 + ⟦ b ⟧ * 2
+    | +-assoc (⟦ a ⟧ * 2) 1 (⟦ b ⟧ * 2)
+    = refl
```

Comment:
Have you tried using the `begin ... ∎` form?

Reply:
Fixed

I'll use begin ... ∎ syntax from now on. It's obviously much easier for communication and much more refactor safe. I just forgot about it and got lazy haha.

Reply:
> It's obviously much easier for communication and much more refactor safe.

I'm glad to know that you think so, too.